src/librbd/cache/ParentCacheObjectDispatch.cc

```diff
@@ -0,0 +1,234 @@
+// vim: ts=8 sw=2 smarttab
+
+#include "common/WorkQueue.h"
+#include "librbd/ImageCtx.h"
+#include "librbd/Journal.h"
+#include "librbd/Utils.h"
+#include "librbd/io/ObjectDispatchSpec.h"
+#include "librbd/io/ObjectDispatcher.h"
+#include "librbd/io/Utils.h"
+#include "librbd/cache/ParentCacheObjectDispatch.h"
+#include "osd/osd_types.h"
+#include "osdc/WritebackHandler.h"
+
+#include <vector>
+
+#define dout_subsys ceph_subsys_rbd
+#undef dout_prefix
+#define dout_prefix *_dout << "librbd::cache::ParentCacheObjectDispatch: " \
+                           << this << " " << __func__ << ": "
+
+using namespace ceph::immutable_obj_cache;
+using librbd::util::data_object_name;
+
+namespace librbd {
+namespace cache {
+
+template <typename I>
+ParentCacheObjectDispatch<I>::ParentCacheObjectDispatch(
+    I* image_ctx) : m_image_ctx(image_ctx), m_cache_client(nullptr),
+    m_object_store(nullptr), m_initialized(false), m_re_connecting(false) {
+  std::string controller_path =
+    ((CephContext*)(m_image_ctx->cct))->_conf.get_val<std::string>("immutable_object_cache_sock");
+  m_cache_client = new CacheClient(controller_path.c_str(), m_image_ctx->cct);
+}
+
+template <typename I>
+ParentCacheObjectDispatch<I>::~ParentCacheObjectDispatch() {
+    delete m_object_store;
+    delete m_cache_client;
+}
+
+template <typename I>
+void ParentCacheObjectDispatch<I>::init() {
+  auto cct = m_image_ctx->cct;
+  ldout(cct, 5) << dendl;
+
+  if (m_image_ctx->parent != nullptr) {
+    ldout(cct, 5) << "child image: skipping" << dendl;
+    return;
+  }
+
+  C_SaferCond* cond = new C_SaferCond();
+  Context* create_session_ctx = new FunctionContext([cond](int ret) {
+    cond->complete(0);
+  });
+
+  create_cache_session(create_session_ctx, false);
+  cond->wait();
+
+  m_image_ctx->io_object_dispatcher->register_object_dispatch(this);
+  m_initialized = true;
+}
+
+template <typename I>
+bool ParentCacheObjectDispatch<I>::read(
+    uint64_t object_no, uint64_t object_off,
+    uint64_t object_len, librados::snap_t snap_id, int op_flags,
+    const ZTracer::Trace &parent_trace, ceph::bufferlist* read_data,
+    io::ExtentMap* extent_map, int* object_dispatch_flags,
+    io::DispatchResult* dispatch_result, Context** on_finish,
+    Context* on_dispatched) {
+  auto cct = m_image_ctx->cct;
+  ldout(cct, 20) << "object_no=" << object_no << " " << object_off << "~"
+                 << object_len << dendl;
+  ceph_assert(m_initialized);
+  string oid = data_object_name(m_image_ctx, object_no);
+
+  /* if RO daemon still don't startup, or RO daemon crash,
+   * or session occur any error, try to re-connect daemon.*/
+  if (!m_cache_client->is_session_work()) {
+    if(!m_re_connecting.load()) {
+      ldout(cct, 20) << "try to re-connct RO daemon. " << dendl;
+      m_re_connecting.store(true);
+
+      Context* on_finish = new FunctionContext([this](int ret) {
+        m_re_connecting.store(false);
+      });
+      create_cache_session(on_finish, true);
```

Comment:
Possible race condition w/ multiple concurrent attempts to re-create the cache session

Reply:
thanks jason. this is a bug. fixed it.